### PR TITLE
WebGLRenderer: Add support for multiple uv sets and transforms.

### LINKED ISSUE #9457: Map repeat affecting other textures' repeat on material
[plunker](https://plnkr.co/edit/UT9rs0Q6TO0gBxwdxLVK?p=preview)

When setting the `map` to repeat the `alphaMap` is also affected.

I'm expecting each face of the cube to have a giant whole in the middle with the edges of the taco time texture showing near the edges of the cube in a 3x4 pattern.

I'm getting the 3x4 pattern of the taco time texture but the `alphaMap`, although it does not repeat it does scale down to the size of the other map. It also produces some nasty streaks.

If you comment out line 35 `material.map.repeat.set(3,4);` you can see what I'm expecting from the `alphaMap`.

##### Three.js version

- [x] Dev
- [x] r79
- [x] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS

### LINKED ISSUE #12608: GLTFLoader does not support texCoord fields
##### Description of the problem 

Today, the Three.js engine does not allow the use of arbitrary UV coordinate sets for textures. Instead, they're hardcoded to channel 0 for all textures, except for lightMap which uses channel 1. As a result, the `texCoord` property of glTF textureInfo objects are ignored. Thus, the glTF loader is not compliant to the spec.

Once the engine supports UV channels better, this issue is a reminder to update the loader too.

##### Three.js version

- [x] Dev
- [ ] r87
- [ ] ...

### PR DESCRIPTION
**Pull request by mrdoob**
Fixes #12608
Fixes #9457

**Description**

Pushing what I have for now. Still trying to figure out how to make this work.

### CODE REVIEW

File: examples/jsm/materials/MeshGouraudMaterial.js [hunk 1]
```diff
@@ -220,7 +218,7 @@ const GouraudShader = {
 		#include <packing>
 		#include <dithering_pars_fragment>
 		#include <color_pars_fragment>
-		#include <uv_pars_fragment>
+		#include <uvs_pars_fragment>
 		#include <uv2_pars_fragment>
```
Comment (LeviPesin) [inline examples/jsm/materials/MeshGouraudMaterial.js:222]: I think you forgot to remove this line.

**Comment by hybridherbst** [comment]
I love that there's progress here! 

Related:
- https://github.com/mrdoob/three.js/issues/12608
- https://github.com/mrdoob/three.js/pull/8278
- https://github.com/google/model-viewer/issues/3515

**Comment by mrdoob** [comment]
Yeah, I'll leave that for another PR though.

**Comment by LeviPesin** [comment]
I can do such PR (cleaning `parameters` and `array` generation in WebGLPrograms and defines generation in WebGLProgram).

**Comment by mrdoob** [comment]
Trying to add multiple texture transforms while I'm at it.

| Before | After |
| --- | --- |
| Screenshot 2023-03-29 at 00 46 25 | Screenshot 2023-03-29 at 00 46 39|

**Comment by mrdoob** [comment]
Managed to get `webgl_loader_gltf_sheen` to renders correctly:

| Before | After |
| --- | --- |
| Screenshot 2023-03-29 at 00 53 39 | Screenshot 2023-03-29 at 00 53 33|

**Comment by mrdoob** [comment]
This GLTF test now passes (if you ignore the seemingly flipped normals one 🤫) :

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/TextureTransformMultiTest

Screenshot 2023-03-29 at 10 32 03

**Comment by mrdoob** [comment]
Figured it out.

Screenshot 2023-03-29 at 14 12 07

**Comment by mrdoob** [comment]
@LeviPesin 

> I can do such PR (cleaning parameters and array generation in WebGLPrograms and defines generation in WebGLProgram).

Keep in mind that this code is the way it is because it has to run as fast as possible. This code is used at init time by *all* the objects in the scene to see if there's a program already compiled that can be used.

If you end up playing with it make sure you do performance tests in slow devices.

File: examples/jsm/loaders/GLTFLoader.js [hunk 2]
```diff
@@ -1814,7 +1814,7 @@ class GLTFTextureTransformExtension {
 
 		if ( transform.texCoord !== undefined ) {
 
-			console.warn( 'THREE.GLTFLoader: Custom UV sets in "' + this.name + '" extension not yet supported.' );
+			texture.uvSet = transform.texCoord;
```
Comment (donmccurdy) [inline examples/jsm/loaders/GLTFLoader.js:1817]: @mrdoob this block will need to be moved below the `texture.clone()` line a few lines down, so we don't modify the texture until after cloning it. The next block should check `texture.uvSet !== transform.texCoord` (not than `!== undefined`) so we don't clone unnecessarily.

Comment (mrdoob) [reply examples/jsm/loaders/GLTFLoader.js:1817]: Why do we have to clone?

File: examples/jsm/loaders/GLTFLoader.js [hunk 3]
```diff
@@ -3122,11 +3122,9 @@ class GLTFParser {
 
 			if ( ! texture ) return null;
 
-			// Materials sample aoMap from UV set 1 and other maps from UV set 0 - this can't be configured
-			// However, we will copy UV set 0 to UV set 1 on demand for aoMap
-			if ( mapDef.texCoord !== undefined && mapDef.texCoord != 0 && ! ( mapName === 'aoMap' && mapDef.texCoord == 1 ) ) {
+			if ( mapDef.texCoord !== undefined ) {
 
-				console.warn( 'THREE.GLTFLoader: Custom UV set ' + mapDef.texCoord + ' for texture ' + mapName + ' not yet supported.' );
+				texture.uvSet = mapDef.texCoord;
```
Comment (donmccurdy) [inline examples/jsm/loaders/GLTFLoader.js:3127]: We'll also potentially need to clone the texture here, if `mapDef.texCoord` is > 0. Probably best to do that conservatively, we can try to reduce cloning later if it's an issue.

File: examples/jsm/loaders/GLTFLoader.js (hunk 2 above)
Comment (donmccurdy) [reply examples/jsm/loaders/GLTFLoader.js:1817]: LittlestTokyo.glb is reusing the same texture in multiple materials, with different `texCoord` values. We're overwriting a shared texture, which causes the `uvSet = 1` on a geometry without additional UVs.

Comment (mrdoob) [reply examples/jsm/loaders/GLTFLoader.js:1817]: Ah I see... Hmm, are you able to modify the PR? I think I'm done.

**Comment by mrdoob** [comment]
Seems like E2E died 🙃

**Comment by mrdoob** [comment]
In theory this PR makes things slower as maps no longer reuse the same `vUv` but I think the flexibility it brings is well worth it.

Another thing we may want to consider is turning `texture.matrixAutoUpdate` to `texture.needsUpdate` and let the developer control it. Right now we're computing the `texture.matrix` of every map used every frame...

**Comment by mrdoob** [comment]
Regarding naming...

@donmccurdy [Here](https://github.com/mrdoob/three.js/issues/12608#issuecomment-381368147) you proposed `texture.channel`. I don't really have any preference, maybe I like `channel` more than `uvSet` because the uppercase `S` is annoying 😇

File: examples/jsm/loaders/GLTFLoader.js (hunk 3 above)
Comment (donmccurdy) [reply examples/jsm/loaders/GLTFLoader.js:3131]: Done: https://github.com/mrdoob/three.js/pull/25721/commits/ef9cee5d5d58cafe68728aa6521517ac74d275b6

File: examples/jsm/loaders/GLTFLoader.js (hunk 2 above)
Comment (donmccurdy) [reply examples/jsm/loaders/GLTFLoader.js]: Done: https://github.com/mrdoob/three.js/pull/25721/commits/ef9cee5d5d58cafe68728aa6521517ac74d275b6

**Comment by donmccurdy** [comment]
> I had to add a check in WebGLPrograms to avoid crashing...

If you'd like, I think the check can be removed after ef9cee5.

I'm OK with any of `channel`, `uvSet`, `uvChannel` or `uvIndex`. Perhaps a slight preference for `.channel` or `.uvChannel`.

**Comment by mrdoob** [comment]
Screenshot 2023-03-30 at 12 35 57

What should it say? 😇

**Comment by pailhead** [comment]
I just got to say, when you read Lorem ipsum first, followed by Italian, brain goes boom. 😂

Thank you, it’s verbose, but it had to be done.

**Comment by pailhead** [comment]
> Screenshot 2023-03-30 at 12 35 57
> 
> What should it say? 😇

“UV channel index. UV channel is a set of two dimensional coordinates that is assigned to the vertices of a model.”

**Comment by mrdoob** [comment]
Links for comparing performance:

http://raw.githack.com/mrdoob/three.js/dev/examples/index.html
http://raw.githack.com/mrdoob/three.js/multi_uvs_builds/examples/index.html

**Comment by mrdoob** [comment]
On a MacBook Air 2015 and on a Pixel 7 I seem to be getting the same FPS.

I feel like it's using more CPU but that's something I'll investigate more deeply and optimize next month (specially `texture.matrixAutoUpdate`).

**Comment by mrdoob** [comment]
Sure does!

https://plnkr.co/edit/Xdnuh8hJDvVUg7VN

Also... Woh! 7 years old code and I only had to add `transparent: true` in the material.

**Comment by Mugen87** [comment]
I've tested some examples with a Pixel 4a and couldn't see FPS differences, too. Although one potentially needs a special test case with many textures to see a potential difference.

**Comment by Mugen87** [comment]
I've noted in the migration guide that the `uv2_*` shader chunks have been removed. I've had a support case where custom shader code relied to it.

**Comment by mrdoob** [comment]
> I've tested some examples with a Pixel 4a and couldn't see FPS differences, too. Although one probably needs a special test case with many textures to see a potential difference.

Yeah, I tried with models with a bunch of maps too and didn't see a difference.

I believe this has been one of the premature optimisations that have damaged the project the most... 😬

Feels pretty good to be outside of that hole now though!

**Comment by mrdoob** [comment]
@donmccurdy @elalish 

Right now `WebGLRenderer` and `GLTFLoader` only supports 2 uv sets.

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLPrograms.js#L38-L44
https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLPrograms.js#L154
https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProgram.js#L530
https://github.com/mrdoob/three.js/blob/dev/examples/jsm/loaders/GLTFLoader.js#L2033-L2034

Adding support 3 or 4 is pretty straight forward though.

Have you ever seen a gltf file with more than 2 uv sets?

**Comment by donmccurdy** [comment]
@mrdoob I've certainly seen glTF files with 3+ UV sets. Usually one or more of the UV sets are unused, I don't know enough about the art workflow to think of cases where 3+ UV sets would be needed. To my understanding Blender, Godot, Unity, and Unreal Engine each support up to 8 UV sets, but various conditions (Shader Graph in Unity, Skinning in Unreal, perhaps more...) will knock that limit down to 4 or fewer. I think supporting 4 UVs would be a nice bit of flexibility, if the performance cost is limited.

@JonathanIcon do you mind explaining what this file does? Is the combination of MSFT_lod and KHR_materials_variants (both contained in the file) important to using additional UV sets here?

**Comment by JonathanIcon** [comment]
@donmccurdy we can ignore the MSFT_LOD yes but we use the KHR_materials_variants having 1 variant per material and each material has a different UV set;

**Comment by pailhead** [comment]
I think good question here would be:

> Can three.js exist without GLTF?

Or

> Is there one, or ten users perhaps, that use three.js without GLTF?

If three can in any scenario exist or be used without GLTF perhaps it shouldn't be the only consideration when designing this feature. Ie. "GLTF has only two channels, let's design three so that it uses only two channels".


If i may suggest an alternative pattern:

```
const g = new BufferGeometry()
g.setAttirbute('uv_0', ...)
g.setAttirbute('uv_1, ...)
g.setAttirbute('uv_2', ...)
g.setAttirbute('uv_3', ...)
```

If we consider these uvs as just array of attributes that exist without GLTF, OBJ, dae, 3ds and whatnot, i think it may become obvious that 2 channels is rather arbitrary :)

**Comment by pailhead** [comment]
This is, is how this editor looks in 3ds max:

image

The `1:map` is basically what three calls `uv` . In practice, this is just the same mesh copied over and over, just like three's attributes. There has to be a limit, usually 32 for stuff like this in max, but in theory a user can add a whole bunch of mapping channels. Maybe even looking up the same texture, but from different UVs.

Feel free to ping me, i've worked with 3d studio max and other sowftware for 20 years before i thought myself how to program.

**Comment by pailhead** [comment]
Hm, it's also worth noting that 3ds max treats all of this as UVW. So in theory i think at any moment you can designate something as UV kinda like:
```
g.setAttribute('uv', g.getAttribute('position'))
```

So in a sense (i don't know much about GLTF) it can be even more generic? I imagine that GLTF saves space and wouldn't write a bunch of zeros for UVW, but write two numbers for UV. Thus probably making something like this incompatible with three.

However:

```
const loader = new GLTFLoader()

loader.load(['position.gltf', 'uv0.gltf', 'uv1.gtlf', 'uv2.gltf', 'uv3.gltf']).then(results=>{
   const g = new BufferGeometry()
   g.setAttribute('position', results[0]...)
   g.setAttribute('uv_0', removeZ( results[1]...)
   ...
})
```

Should be pretty generic if applying the same pattern. Any mesh, as long as it's the same can be combined into a geometry?

**Comment by pailhead** [comment]
>@mrdoob I've certainly seen glTF files with 3+ UV sets. Usually one or more of the UV sets are unused, I don't know enough about the art workflow to think of cases where

The art flow from an engineering point can be seen as very generic. It's a creative process, and there may be a way to use several UVs while reusing several textures to create an effect that no one has thought of. We are limited by the number of attributes in GL land, but when considering that a technical artist can be involved, i wouldn't find it unreasonable for someone to have a need to use say 8 channels, while using no attributes for skinning or perhaps even normals. 

I think the channel info tool from 3ds max is a good example.

**Comment by donmccurdy** [comment]
Many devices are limited to 16 vertex attributes. In theory we could support up to 15 UVs + 1 position attribute, but in practice it may reduce complexity to have some limit. We are not designing around glTF here — it has no limit either — just trying to compare notes on use cases, and how often we see them.

**Comment by pailhead** [comment]
Correct, so maybe even something like:

```
const g = new BufferGeometry()
g.addUVChannel( new BufferAttribute(...) )
g.addUVChannel( new BufferAttribute(...) )
```
This wires it to three's `uv` `uv2` or whatever is the representation, whereas maybe it wouldnt make sense, but would be allowed to do
```
const g = new BufferGeometry()
const channel = new BufferAttribute(...)
g.addUVChannel(  channel )
g.setAttribute('horse', channel)
```

**Comment by mrdoob** [comment]
#25788

**Pull request merged**